AIP-67 - Multi-team: Verify dag/task executors are present in team

o-nikolas opened the pull request:
If a dag or task is configured to use a specific executor (leveraging Multiple Executor Configuration) and the dag belongs to a team, then validate that the executor is present in that team before adding the dag to the dag bag.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

vincbeck commented on airflow-core/src/airflow/models/dagbag.py:
```diff
@@ -147,12 +147,47 @@ def handle_timeout(signum, frame):
+                    "Found team '%s' for DAG '%s' via bundle '%s'", dag_team_name, dag.dag_id, dag.bundle_name
+                )
+            else:
+                log.debug("Bundle '%s' not found in configuration for DAG '%s'", dag.bundle_name, dag.dag_id)
```
This branch should not happen right? Or if so, we should raise an exception. Maybe a method in `DagBundlesManager` like `get_config` which returns the config of a dag bundle given its name, otherwise raise an error

o-nikolas replied:
Yeah, I suppose it was a bit cautious. I can just delete this branch altogether, we default to None for the team any way.

o-nikolas replied:
Updated

vincbeck commented on airflow-core/src/airflow/models/dagbag.py:
```diff
@@ -147,12 +147,44 @@ def handle_timeout(signum, frame):
+                "Found team '%s' for DAG '%s' via bundle '%s'", dag_team_name, dag.dag_id, dag.bundle_name
+            )
+        except Exception as e:
+            log.debug(
```
I am trying to understand in which scenario we end up in this branch. Can you provide an example? My feeling tells me this is a anormal situation, thus we should raise an exception, not swallow it

o-nikolas replied:
Yupp, see the ToDo above. For now I do not want this code to fail the dag parsing in ANY possible way, so all exceptions are caught and logged. As we test, tweak and gain comfort with this approach this debug statement will be changed to an exception being raised (or the try/catch removed and the exceptions allowed to bubble up).

vincbeck replied:
Sounds good

o-nikolas replied:
I will likely switch this to exception once we have the multi-team flag from your PR @vincbeck I think that will be a great compromise!

o-nikolas committed:
Verify dag/task executors are present in team

If a dag or task is configured to use a specific executor (leveraging
Multiple Executor Configuration) and the dag belongs to a team, then
validate that the executor is present in that team before adding the dag
to the dag bag.

jedcunningham commented on airflow-core/src/airflow/dag_processing/dagbag.py:
```diff
@@ -135,12 +135,44 @@ def handle_timeout(signum, frame):
+            bundle_config = bundle_manager._bundle_config[dag.bundle_name]
+            # TODO[multi-team] Raise exceptions below instead of logging once we have a multi-team feature flag configuration
+            dag_team_name = bundle_config.team_name
+            log.debug(
```
Should we only log if we actual found a team name?

o-nikolas replied:
Fixed!

jedcunningham commented on airflow-core/tests/unit/models/test_dagbag.py:
```diff
@@ -83,6 +83,102 @@ def test_validate_executor_field():
+
+        with DAG("test-dag", schedule=None) as dag:
+            # Simulate DAG having bundle_name set during parsing
+            dag.bundle_name = "test_bundle"
```
Is this actually set on dag during parsing though? I didn't think it was.

o-nikolas replied:
I'll do some more testing with this. I've updated the code to be wrapped in the multi-team feature flag. So it'll be safe to merge and experiment further with.

o-nikolas committed:
Add multi_team config gate

o-nikolas committed:
Update tests after move

o-nikolas committed:
Fix tests with new config gate
